### ISSUE TITLE
serverless-offline raises warning due to callback usage

### ISSUE THREAD

**Issue by ceilfors** [laconiajs/laconia#48]
**Describe the bug**
This bug is originally raised by @Strernd on gitter. When Laconia is used together with serverless-offline, it is raising a warning.

**To Reproduce**
Use Laconia with serverless-offline

**Expected behavior**

**Actual behavior**
A warning message generated by serverless-offline:

```
Serverless: Warning: handler 'laconia-handler' returned a promise and also uses a callback! | This is problematic and might cause issues in your lambda.
```

**Additional context**
The callback is originally retained even though we support node 8 because of a problem in AWS Lambda. Seems like this problem is now fixed, see: https://stackoverflow.com/a/51988695/2464295

**Comment by ceilfors** [comment]
Removing the use of callback will improve the ease of catching errors in middlewares too, as at the moment users who are trying to catch errors in middleware will have to implement a callback.

**Comment by ceilfors** [comment]
Released under 1.4.0

**Comment by ceilfors** [comment]
Released under 1.4.0.

**Comment by ceilfors** [comment]
@geoffdutton Upon running the `acceptance-test` in laconia, this change of code apparently is breaking `@laconia/batch` behaviour. The break has nothing to do with our code, but AWS Lambda itself. Without the use of callback, AWS Lambda doesn't seem to be waiting for the event loop to be empty before ending its execution, hence the self invocation mechanism (recursion) that's supposed to happen in `@laconia/batch` is not happening.

This problem is similar to the behaviour that has been described here: https://forums.aws.amazon.com/thread.jspa?messageID=871017

**Comment by ceilfors** [comment]
Unfortunately the advice from that forum doesn't seem to work. I have attempted to remove the `async` from top level, by changing the code to look like this

```js
  const laconia = (event, context) => {
    laconiaContext.registerInstances({ event, context });

    return laconiaContext.refresh().then(() => app(event, laconiaContext));
  };
```

And it doesn't seem to still be waiting for the event loop to be empty.

Reverting the code has definitely make the `acceptance-test` to pass. Any other alternatives before we attempt to revert the code?

**Comment by geoffdutton** [comment]
I was trying to run the acceptance tests as well and couldn't get them to work. This makes sense. What if we make the main `laconia` function something like:

```
// This would be the raw Lambda event, context, and callback
const laconia = (event, context, callback) => {
    laconiaContext.registerInstances({ event, context });

    laconiaContext.refresh()
        .then(() => app(event, laconiaContext));
        .then(res => {
            callback(null, res)
        })
        .catch(err => {
            callback(err)
        })
};
```

I think if we don't return a promise and only use the call back at the very top level, maybe it works? I'll try to test it tonight or tomorrow.

**Comment by ceilfors** [comment]
On running the acceptance test, do leave your issues or feedback in #104 or #99.

Yes, this is the opposite problem of callbackWaitsForEmptyEventLoop, and I presume it's a bug.. The suggestion you made works as the acceptance test is passing! Just had to add a return statement.

Even though it works, I'm quite torn apart with the use of callback. One of the big win in term of design when `callback` is not used is, when a middleware is being written, they will not have to handle any callback at all. They will be able to handle error flow just by using try catch.

I had a use case where I wanted to have a highest level middleware where I handle errors differently, but unfortunately I had to pass callback all the way down, and check if the callback is called with an error object. The removal of callback will make this use case super easy.

I wonder if there's a way to call the callback on the very last minute or something, maybe by calling setImmediate. I'll experiment tonight.

**Comment by ceilfors** [comment]
@geoffdutton I've reverted the change because I just realised that the use of `.then` works because if added a `return` statement. Upon removing the `return` statement, the test breaks which means other changes are necessary.

To help release our other changes, I'll revert this for now

**Comment by hugosenari** [comment]
we could at least ignore callback if it does not exist :thinking:

**Comment by hugosenari** [comment]
It is a breaking change since someone could wrap laconia/core with something like:

```javascript
cons lac = laconia(() => true);

cons handler = (event, context, callback) => {
  const myCallback = (...args) => {
    console.log(args);
    callback(...args);
  };
  lac(event, context, myCallback);
};
```

`myCallback` will never be called in our newer version.

Options:
1. Update major version;
2. Create another entrypoint (ie: @laconia/core/async)

**Comment by ceilfors** [comment]
@hugosenari Good point. Another option is, we could introduce a configuration that we eventually deprecate i.e. LACONIA_NO_CALLBACK = true.

**Comment by hugosenari** [comment]
Fixed by #511 (v1.9.0)
Now we can add LACONIA_NO_CALLBACK='true' to env or omitting callback parameter.
